### PR TITLE
[pipelining] Support arbitrary stage ordering on ranks

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129810
* #129786
* #129264
* #129233
* __->__ #128976
* #128983

Fixes based on discussion in https://github.com/pytorch/pytorch/issues/128665

Our previous assumption was that for looped schedules `stage_ids = range(rank, total_stages, num_local_stages)`. This is not true for all schedules. This change relaxes that assumptions and allows arbitrary ordering of stages. For example in the added test we do, rank 0: [stage0, stage3], rank 1: [stage1, stage2]. The test also adds a schedule registry (for testing) which performs 1 microbatch through this schedule

```
F0_0 None None F0_3 B0_3 None None B0_0
None F0_1 F0_2 None None B0_2 B0_1 None
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: test/distributed/pipelining/schedule_registry.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+from typing import Any, Callable, Dict, List, Optional
+
+import torch
+from torch.distributed.pipelining.schedules import _ComputationType, _Action, PipelineScheduleMulti
+from torch.distributed.pipelining.stage import _PipelineStageBase
+
+F = _ComputationType.FORWARD
+B = _ComputationType.BACKWARD
+
+class ScheduleUnevenLooped(PipelineScheduleMulti):
```
Comment (wconstab) [inline test/distributed/pipelining/schedule_registry.py:14]: nit: it's not really looped, its V, isn't it?

I like the idea of the schedule registry

File: torch/distributed/pipelining/schedules.py [hunk 2]
```diff
@@ -656,12 +663,19 @@ def _step_microbatches(
         stage_index_to_stage: Dict[int, _PipelineStageBase] = {
             stage.stage_index: stage for stage in self._stages
         }
-        prev_rank: int = (self.rank - 1) % self.pp_group_size
-        next_rank: int = (self.rank + 1) % self.pp_group_size
+
+        # determine prev_rank and next_rank based on which ranks are next to
+        # the stages in the pipeline_order
+        all_prev_ranks: Set[int] = set()
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:669]: the naming (all) suggests more than one next rank, which is a little confusing.

i think the intent is that later we could add skip connection support?
- if we added it, i'd want to rename perhaps to 'fwd_send_ranks' rather than using the word 'next' which implies linearity together with 'all' which applies multiple

File: torch/distributed/pipelining/schedules.py [hunk 3]
```diff
@@ -656,12 +663,19 @@ def _step_microbatches(
...
+        all_next_ranks: Set[int] = set()
+        for stage_index in stage_index_to_stage.keys():
+            # TODO: assumption that stages only communicate from distances of +1/-1 (no skip connections)
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:672]: you're being a bit uncommitted i think- should we generalize the code to support skip connections?  

if so we need tests for that, and we need a solution to this issue - how to determine who the other ranks are at this point in the code?

or if not, maybe its not great to partialy generalize the code below (making for-loops e.g. `for prev_rank in all_prev_ranks:`

perhaps better to fully punt on 'skip connections' till we can fully test it and ensure its free of edge cases etc.  and only land the smaller set of changes needed to enable decoupling rank from stage id.

File: torch/distributed/pipelining/schedules.py [hunk 4]
```diff
@@ -565,6 +565,7 @@ def __init__(
         args_chunk_spec: Optional[Tuple[TensorChunkSpec, ...]] = None,
         kwargs_chunk_spec: Optional[Dict[str, TensorChunkSpec]] = None,
         output_merge_spec: Optional[Union[Dict[str, Any], Tuple[Any]]] = None,
+        stage_index_to_group_rank: Optional[Dict[int, int]] = None,
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:568]: Perhaps we should create 2 classes for 'PipelineScheduleLooped' and 'PipelineScheduleV' that provide the right default mapping. It wouldn't preclude a user creating a new type of mapping later but it would be a convenience and cover the known cases

File: torch/distributed/pipelining/schedules.py [hunk 5]
```diff
@@ -583,6 +584,12 @@ def __init__(
         self._num_stages = stages[0].num_stages
         self.pp_group_size = stages[0].group_size
         self.rank = stages[0].group_rank
+        # Set the pipeline stage states
+        if stage_index_to_group_rank is not None:
+            for stage in self._stages:
+                stage.stage_index_to_group_rank = stage_index_to_group_rank
+        self.stage_index_to_group_rank = stages[0].stage_index_to_group_rank
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:591]: its a little weird to me to have the schedule set this inside the stage.  Does the stage need this information for creating ops? 

would it be cleaner if the schedule passed the 'dest rank' as an integer to stage.get_fwd_send_ops, for example?

File: torch/distributed/pipelining/schedules.py (hunk 2 above)
Comment (H-Huang) [reply torch/distributed/pipelining/schedules.py:669]: This will also help to support skip connections. But this is still needed if we truly want to support arbitrary stage per rank since there could be multiple prev ranks. For example:

rank0: [stage0, stage2]
rank1: [stage1, stage4]
rank2: [stage2, stage3]

rank1 have a previous rank0 (from stage0) and previous rank2 (from stage3). This is a contrived example, but this list containing multiple ranks is needed if we want to have no restrictions for the user.

File: test/distributed/pipelining/schedule_registry.py (hunk 1 above)
Comment (H-Huang) [reply test/distributed/pipelining/schedule_registry.py:14]: Oops yeah, will update the name!

File: torch/distributed/pipelining/schedules.py (hunk 3 above)
Comment (H-Huang) [reply torch/distributed/pipelining/schedules.py:672]: I wanted to punt on skip connections as well, but my hope is that if the changes in this PR are landed then it should be easy to add as we really need only need to update the stage API for users to pass in output arg to stage mappings, and then update this area of the code you commented on.

The for-loop is necessary regardless of skip connections due to the reason mentioned in the previous comment.

Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:672]: wait.. is the for loop even being tested? my read of this was that all_prev_ranks starts out empty, and we just add one rank to it which is the rnak for stage_index-1.

where do we add a second rank to all_prev_ranks?

if its just one rank, then the for loop is just running the same way as it used to when it was not a for loop isn't it?

File: torch/distributed/pipelining/schedules.py [hunk 6]
```diff
@@ -684,43 +698,47 @@ def _step_microbatches(
...
-                elif computation_type == _ComputationType.BACKWARD:
-                    # If not the first stage, then receive bwd gradients
-                    if stage_index != 0:
-                        # TODO: We are assuming that stage will always receive from stage+1
-                        # however that is not necessarily true of get_bwd_recv_ops
-                        stage = stage_index_to_stage[stage_index - 1]
-                        ops.extend(stage.get_bwd_recv_ops(mb_index))
-                else:
-                    raise ValueError(f"Unknown computation type {computation_type}")
+            for prev_rank in all_prev_ranks:
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:701]: iiuc, all_prev_ranks for rank 1 would include 
 - rank 0, which is prev in the forward pass
 - rank 2, which is prev in the backward pass

I'm confused in general by the logic here.  i guess i would expect that we have a special case for when 'bwd_prev_rank' is doing a bwd step, then we need to recv from it.  and another special case for when 'fwd_prev_rank' is doing forward, then we need to recv forward input from it.

but it looks like we mix together both prev_ranks into one for loop and then inside the loop we ignore the case where the prev rank is doing backward. 

so i take this to mean that the variable 'all_prev_ranks' really refers to 'prev_forward_rank'?

File: torch/distributed/pipelining/schedules.py (hunk 3 above)
Comment (H-Huang) [reply torch/distributed/pipelining/schedules.py:672]: In the test case the rank to stages are: 

rank0: stage0, stage3
rank1: stage1, stage 2

Since we just check stage_index-1, for rank 1 all_prev_rank would be [0, 1] (it includes itself), so the loop does execute.

File: torch/distributed/pipelining/schedules.py (hunk 4 above)
Comment (H-Huang) [reply torch/distributed/pipelining/schedules.py:568]: I think I would be okay with this, so then PipelineScheduleLooped always ensures that `rank = stage_index % group_size` and we don't allow the `stage_index_to_group_rank` arg?

I believe it should be a name other than`PipelineScheduleV` since the `stage_index_to_group_rank` provides any arbitrary stage ordering, not necessarily just a "V-shape"

File: torch/distributed/pipelining/schedules.py (hunk 5 above)
Comment (H-Huang) [reply torch/distributed/pipelining/schedules.py:591]: The stage does need this information since when it creates the `RecvInfo` classes to store the activations / grads (https://github.com/pytorch/pytorch/blob/main/torch/distributed/pipelining/stage.py#L52), it only stores the stage index.

When we actually create the ops it converts from `stage_index` -> `group_rank` (uses mapping) -> `global_rank` (uses distributed_c10d)

File: torch/distributed/pipelining/schedules.py (hunk 6 above)
Comment (H-Huang) [reply torch/distributed/pipelining/schedules.py:701]: Yeah 'prev_forward_ranks' is a more accurate name. My goal for the logic was, in this current step:

1) look for any forward activations I need from any ranks in the group (`prev_forward_ranks`)
2) look for any backwards gradients I need from any ranks in the groups.

Maybe a clearer name might be something like `sending_activation_ranks` (for the forward ranks) and `sending_gradient_ranks` (for the backward ranks) instead

File: torch/distributed/pipelining/schedules.py (hunk 3 above)
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:672]: i think i was confused that we were setting 'all_prev_ranks' for multiple stages on this rank at the same time.  i missed your change above for adding a for loop over stage id.

i think this might actually be a risk.  let me add a comment there

File: torch/distributed/pipelining/schedules.py [hunk 7]
```diff
@@ -656,12 +663,19 @@ def _step_microbatches(
...
             stage.stage_index: stage for stage in self._stages
         }
-        prev_rank: int = (self.rank - 1) % self.pp_group_size
-        next_rank: int = (self.rank + 1) % self.pp_group_size
+
+        # determine prev_rank and next_rank based on which ranks are next to
+        # the stages in the pipeline_order
+        all_prev_ranks: Set[int] = set()
+        all_next_ranks: Set[int] = set()
+        for stage_index in stage_index_to_stage.keys():
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:671]: is it safe to loop over stages here?

R0 [S0, S7]
R1 [S1, S6]
R2 [S2, S5]
R3 [S3, S4]

For rank 2- what if S2 needs to send forward to S3 for a later mb_chunk and S5 needs to recv forward from S4 for an earlier chunk?  This is just one attempt to construct an example where 2 pairs of stages are communicating across rank boundary at the same timestep.

Do we need to batch comms for S2-S3 with comms for S4-S5?

IIUC the current for loop separates batching to stage-boundary.  I think batching only needs to be separated by rank boundary, and may be required to be batched across stage-boundary in some cases.  Wdyt?

File: torch/distributed/pipelining/schedules.py (hunk 6 above)
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:701]: see my other comment (https://github.com/pytorch/pytorch/pull/128976/files#r1647927426) - i wonder if it would be cleaner/safer to 
- first list out the prev-rank and the next-rank, (assuming no skip conns)
- then find any send/recv between self and prev-rank (for either combination of local stages forward or backward) and batch those comms
- same for next-rank

File: torch/distributed/pipelining/schedules.py (hunk 7 above)
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:671]: ok, i'm wrong, sorry. I see that you are building a list of ops and then later executing the whole list. So you are already batching across stages, it doesn't really matter about the for loop.

Comment (H-Huang) [reply torch/distributed/pipelining/schedules.py:671]: In the current runtime, there can only be a single action per timestep, which means that only 1 stage / 1 mb per rank can operate at any point in time.

So the case with S2 -> S3 and S5 -> S4 wouldn't happen in the same timestep. This change doesn't break that assumption it just changes the number of ranks that any rank can communicate with.

Previously a rank could only communicate with 0-2 other ranks (forward/backward), now a rank can potentially communicate with all ranks.

File: torch/distributed/pipelining/schedules.py [hunk 8]
```diff
@@ -684,43 +698,47 @@ def _step_microbatches(
...
+                        if stage_index - 1 in stage_index_to_stage:
+                            # TODO: We are assuming that stage will always receive from stage+1
+                            # however that is not necessarily true of get_bwd_recv_ops
+                            stage = stage_index_to_stage[stage_index - 1]
+                            ops.extend(stage.get_bwd_recv_ops(mb_index))
+                    else:
+                        raise ValueError(f"Unknown computation type {computation_type}")
 
             # do the communication
             if ops:
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:744]: something unrelated to the current PR but might be an issue with the runtime approach we have taken-

we batch all the comm ops for the current timestep, and then issue them after the current action.  Then we come back to do another action.  It looks like we would not have any opportunity to overlap comms.

Specifically, we seem to have this pattern

Action()
Comms().wait()
Action()
comms().wait()

Which IIUC would result in a cuda timeline like 
[Action][Comms][Action][Comms]...

What if we want to do some of the comms so they overlap with some of the actions?

**Review by wconstab — APPROVED** [review]
stamp to unblock. needs future work to address .wait() in the right place for overlap.
